src/transform/src/predicate_pushdown.rs

```diff
@@ -179,25 +195,69 @@ impl PredicatePushdown {
                         equivalences,
                         ..
                     } => {
-                        // We want to scan `predicates` for any that can apply
-                        // to individual elements of `inputs`.
+                        // We want to scan `predicates` for any that can
+                        // 1) become join variable constraints
+                        // 2) apply to individual elements of `inputs`.
+                        // Figuring out the set of predicates that belong to
+                        //    the latter group requires 1) knowing which predicates
+                        //    are in the former group and 2) that the variable
+                        //    constraints be in canonical form.
+                        // Thus, there is a first scan across `predicates` to
+                        //    populate the join variable constraints
+                        //    and a second scan across the remaining predicates
+                        //    to see which ones can become individual elements of
+                        //    `inputs`.
 
                         let input_mapper = expr::JoinInputMapper::new(inputs);
 
-                        // Predicates to push at each input, and to retain.
+                        // Predicates not translated into join variable
+                        // constraints
+                        let mut pred_not_translated = Vec::new();
+
+                        for predicate in predicates.drain(..) {
+                            // Translate `col1 == col2` constraints into join variable constraints.
+                            use expr::BinaryFunc;
+                            use expr::UnaryFunc;
+                            if let MirScalarExpr::CallBinary {
+                                func: BinaryFunc::Eq,
+                                expr1,
+                                expr2,
+                            } = &predicate
+                            {
+                                if !expr1.is_literal() && !expr2.is_literal() {
+                                    // An equality predicate implies that both
+                                    // expressions are not null.
+                                    // Since we try to push non-equality predicates
+                                    // at all inputs, we only need to push one
+                                    // predicate into `pred_not_translated` for it
+                                    // to be pushed at both expressions.
+                                    pred_not_translated.push(
+                                        expr1
+                                            .clone()
+                                            .call_unary(UnaryFunc::IsNull)
+                                            .call_unary(UnaryFunc::Not),
+                                    );
+                                    equivalences.push(vec![(**expr1).clone(), (**expr2).clone()]);
```

Comment:
@frankmcsherry I'm confused about what you mean by the atomicity problem since the `!(isnull(..))` is introduced before pushing the expression into the equivalences.

Reply:
Me too. I looked at some code that didn't look like the above, but I def agree that this looks fine. I suspect it's a matter that I didn't understand the `Join` logic where we appear to also form and push down these predicates. Please ignore any "atomicity" concerns I have raise, unless I remember what I was getting at and provide clearer explanation!

Reply:
Re-reading: I think part of my confusion is that the comments for the above expression building leap a few steps forward, and I missed that what the line actually does is simpler. Independent of pushing any expressions down at either input, the line above just adds the predicate to the list of "predicates that still need to hold". We may push it down later, as the comments say, but it could instead just be stated that we add this predicate to filter out matches on null keys.

Now, this was my confusion, certainly. It's resolved now, but if we can insure against future confusions that would help!